### PR TITLE
feat(api): Add AsyncThrowingSequence API.subscribe

### PR DESCRIPTION
**Pull request by lawmicha**
*Issue #, if available:*

*Description of changes:*
This PR introduces the AsyncThrowingSequence version of API.subscribe.
```swift
func subscribe<R>(request: GraphQLRequest<R>) -> AmplifyAsyncThrowingSequence<GraphQLSubscriptionEvent<R>>
```

 I also took the opportunity to rename the subscription event that is received from `SubscriptionEvent<GraphQLResponse<T>>` to `GraphQLSubscriptionEvent<T>`. A subscription event is always going to be a GraphQLResponse on the `data` case, so there's no need for a nested generic. 

*Check points: (check or cross out if not relevant)*

- [ ] Added new tests to cover change, if needed
- [ ] Build succeeds with all target using Swift Package Manager
- [ ] All unit tests pass
- [ ] All integration tests pass
- [ ] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)
- [ ] Documentation update for the change if required
- [ ] PR title conforms to conventional commit style
- [ ] If breaking change, documentation/changelog update with migration instructions

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/GraphQLModelBasedTests+Concurrency.swift [hunk 1]
```diff
@@ -11,75 +11,71 @@ import XCTest
...
-                            connectedInvoked.fulfill()
-                        case .disconnected:
-                            disconnectedInvoked.fulfill()
-                        }
-                    case .data(let result):
-                        switch result {
-                        case .success(let post):
-                            if post.id == uuid {
-                                progressInvoked.fulfill()
+        let sequences = await withTaskGroup(of: AmplifyAsyncThrowingSequence<GraphQLSubscriptionEvent<Post>>.self) { group -> [AmplifyAsyncThrowingSequence<GraphQLSubscriptionEvent<Post>>] in
```
Comment (lawmicha) [inline AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/GraphQLModelBasedTests+Concurrency.swift:26]: This "concurrent subscription" test has been fixed after migrating the code to use TaskGroups, cc @brennanMKE

File: AmplifyPlugins/API/Sources/AWSAPIPlugin/Operation/AWSGraphQLSubscriptionTaskRunner.swift [hunk 2]
```diff
@@ -10,6 +10,177 @@ import Foundation
...
+
+    private func onGraphQLResponseData(_ graphQLResponseData: Data) {
+        do {
+            let graphQLResponseDecoder = GraphQLResponseDecoder(request: request, response: graphQLResponseData)
+            let graphQLResponse = try graphQLResponseDecoder.decodeToGraphQLResponse()
+            send(.data(graphQLResponse))
+        } catch let error as APIError {
+            fail(error)
+        } catch {
+            // TODO: Verify with the team that terminating a subscription after failing to decode/cast one
```
Comment (brennanMKE) [inline AmplifyPlugins/API/Sources/AWSAPIPlugin/Operation/AWSGraphQLSubscriptionTaskRunner.swift:154]: Please add a subtask to create a unit test for this scenario.

File: AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/GraphQLModelBasedTests+Concurrency.swift (hunk 1 above)
Comment (brennanMKE) [reply AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/GraphQLModelBasedTests+Concurrency.swift:26]: 👍

**Review by brennanMKE — DISMISSED** [review]
These tests are looking better all the time. Well done. 👍

File: AmplifyPlugins/API/Sources/AWSAPIPlugin/Operation/AWSGraphQLSubscriptionTaskRunner.swift (hunk 2 above)
Comment (lawmicha) [reply AmplifyPlugins/API/Sources/AWSAPIPlugin/Operation/AWSGraphQLSubscriptionTaskRunner.swift:154]: Added subtask for internal tracking

**Review by brennanMKE — DISMISSED** [review]

**Review by brennanMKE — APPROVED** [review]

**Pull request merged**